File: app/src/main/java/com/wwxiaoqi/wechat_backup/thread/BackupThread.java
package com.wwxiaoqi.wechat_backup.thread;

import android.app.ProgressDialog;
import android.content.Context;
import com.wwxiaoqi.wechat_backup.R;
import com.wwxiaoqi.wechat_backup.utils.ShellUtils;
import java.io.File;

public class BackupThread extends Thread {
  private File fileDir;
  private Context context;
  private ProgressDialog progressDialog;

  public BackupThread(Context context, File fileDir) {
    this.fileDir = fileDir;
    this.context = context;
    progressDialog = new ProgressDialog(context);
    progressDialog.setMessage(context.getString(R.string.backup_prompt));
    progressDialog.setCancelable(false);
    progressDialog.show();
  }

  @Override
  public void run() {
    try {
      sleep(3000);
      new Thread(() -> {
        String[] cmd = { "su", "source " + fileDir + "/backup.sh" };
        ShellUtils.execCmd(cmd, true);
      }).start();
      progressDialog.setMessage(context.getString(R.string.backup_success_prompt));
      sleep(3000);
      progressDialog.dismiss();
    } catch (InterruptedException e) {
      e.printStackTrace();
    }
  }

}

File: app/src/main/java/com/wwxiaoqi/wechat_backup/thread/BackupThread.kt
package com.wwxiaoqi.wechat_backup.thread

import android.os.Handler
import android.os.Message
import com.wwxiaoqi.wechat_backup.utils.ShellUtils
import java.io.File

class BackupThread(
  private val fileDir: File,
  private val timeHandler: Handler
) : Thread() {
  override fun run() {
    try {
      Thread {
        val cmd = arrayOf("su", "source $fileDir/backup.sh")
        ShellUtils.execCmd(cmd, true)
      }.start()
      sleep(3000)
      val message: Message = Message.obtain()
      message.what = 0
      timeHandler.sendMessage(message)
    } catch (e: InterruptedException) {
      e.printStackTrace()
    }
  }
}